src/rt/lifetime.d

```diff
@@ -1193,24 +1186,53 @@ extern (C) void _d_delarray_t(void[]* p, const TypeInfo_Struct ti)
 {
     if (p)
     {
-        assert(!(*p).length || (*p).ptr);
-        if ((*p).ptr)
+        auto bic = __getBlkInfo(p.ptr);
+        auto info = bic ? *bic : GC.query(p.ptr);
+
+        if (info.base && (info.attr & BlkAttr.APPENDABLE))
```

Comment:
Not sure, but I'd tend to check APPENDABLE too, because of this.

```d
class Foo
{
    static struct Bar { ~this() {} }
    Bar[4] bars;
}

unittest
{
    auto foo = new Foo;
    auto bars = foo.bars[];
    delete bars;
}
```